OpenSim/Simulation/Test/testStatesTrajectory.cpp

```diff
@@ -508,6 +508,386 @@ void testCopying() {
+        SimTK_TEST(states.findIndexAt(0.49999, 1.0e-5 + SignificantReal) == 0);
+        SimTK_TEST(states.findIndexAt(1.299999, 1.0e-6 + SignificantReal) == 1);
+        SimTK_TEST(states.findIndexAt(1.300001, 1e-6) == 1);
+        SimTK_TEST(states.findIndexAt(3.499, 1e-3) == 2);
+        SimTK_TEST(states.findIndexAt(3.501, 1e-3) == 2);
+        SimTK_TEST(states.findIndexAt(4.09999999, 1e-8) == 3);
+        SimTK_TEST(states.findIndexAt(4.10000001, 1e-8) == 3);
+
+        // Asking for a state that doesn't exist, even with the tolerance.
+        SimTK_TEST_MUST_THROW_EXC(
+                states.findIndexAt(3.6 + SimTK::SignificantReal, 0.1),
+                StatesTrajectory::NoStateAtTime);
+        SimTK_TEST_MUST_THROW_EXC(
+                states.findIndexAt(4.12 + SimTK::SignificantReal, 0.02),
+                StatesTrajectory::NoStateAtTime);
+    }
+
+    // Make sure that you get the closest state if multiple states are within
+    // the tolerance.
+    {
+        StatesTrajectory states;
+        state.setTime(8.03);
+        states.append(state); // 0
+        state.setTime(8.04);
+        states.append(state); // 1
+        state.setTime(8.05);
+        states.append(state); // 2
+        state.setTime(8.06);
+        states.append(state); // 3
+
+        // Both the 8.04 and 8.05 states are within the tolerance for the
+        // requested times below, but we expect to get the closer of the two.
+        SimTK_TEST(states.findIndexAt(8.044, 0.01) == 1);
+        SimTK_TEST(states.findIndexAt(8.046, 0.01) == 2);
+
+        SimTK_TEST(states.findIndexAt(8.04499999999, 0.01) == 1);
+        SimTK_TEST(states.findIndexAt(8.04500000001, 0.01) == 2);
+
+        // It actually doesn't matter what the tolerance is too much; the
+        // tolerance can even extend beyond the range of times in trajectory.
+        SimTK_TEST(states.findIndexAt(8.044, 5.0) == 1);
+        SimTK_TEST(states.findIndexAt(8.046, 5.0) == 2);
+    }
+
+    // Make sure that you get the closest state if multiple states are within
+    // the tolerance, even at the same time.
+    {
+        StatesTrajectory states;
+        state.setTime(7.2);
+        states.append(state); // 0
+        state.setTime(7.5);
+        states.append(state); // 1
+        state.setTime(7.5);
+        states.append(state); // 2
+        state.setTime(7.5);
+        states.append(state); // 3
+        state.setTime(7.9);
+        states.append(state); // 4
+        state.setTime(7.9);
+        states.append(state); // 5
+        state.setTime(8.0);
+        states.append(state); // 6
+
+        // If you ask for a time at which there are multiple states (with
+        // the same exact time), you get the first of these states.
+        SimTK_TEST(states.findIndexAt(7.5) == 1);
+        SimTK_TEST(states.findIndexAt(7.9) == 4);
+        SimTK_TEST(states.findIndexAt(7.4, 0.1) == 1);
+        SimTK_TEST(states.findIndexAt(7.56, 0.06) == 3);
```

Comment:
Not sure what you mean by:

> you get the first of these states
index 3 is the third state with time 7.5 not the first (1)